src/System.Private.ServiceModel/tests/Stress/SharedPoolsOfWCFObjects/CallStats.cs

```diff
@@ -0,0 +1,205 @@
+                SunnyDay.AddCallTiming(sw.ElapsedTicks);
+            }
+            catch (Exception e)
+            {
+                RainyDay.AddCallTiming(sw.ElapsedTicks);
+            }
+        }
+    }
+
+    public class TimingPercentileStats
+    {
+        public DateTime Time;       // Time when the stats were calculated
+        public int[] Centile;       // negative % means take bottom part, positive % means take top part
+        public long[] AvgTiming;    // within this time
+    }
+
+    public class CallTimingStatsCollector
+    {
+        private int _samples;
+        private int _numCalls;
+        private long[] _timings;
+        private Object _lock = new Object();
+        private Func<long[], TimingPercentileStats> _statsCalculator;
+        public ConcurrentQueue<TimingPercentileStats> PercentileStats { get; set; }
+
+        public CallTimingStatsCollector(int samples, Func<long[], TimingPercentileStats> statsCalculator)
+        {
+            _samples = samples;
+            _timings = new long[_samples * 2];
+            _statsCalculator = statsCalculator;
+            PercentileStats = new ConcurrentQueue<TimingPercentileStats>();
+        }
+
+        // For the most part lock-free and isn't particularly thread safe
+        //
+        // However here we use it without any additional synchronization 
+        // since we use a large number of samples and 
+        public void AddCallTiming(long ticks)
+        {
+            int index = Interlocked.Increment(ref _numCalls) % (_samples * 2);
+            _timings[index] = ticks;
+
+            if (index % _samples == 0)
+            {
+                // if we get a context switch here then the data in _timings may get overwritten by other threads calling AddCallTiming()
+                // however no data will be lost if there were fewer than _samples calls while we make a copy of our half of _timings
+                // the copy may land into LOH - before trying to optimize it we'll need to optimize _statsCalculator as well
+                long[] copy = new long[_samples];
+                int timingsIndex = index == 0 ? _samples + 1 : 1;
+                // this almost useless lock will only stop the first thread that wrapped up _samples calls while we copy the data
+                lock (_lock)
+                {
+                    for (int i = 0; i < _samples; i++)
+                    {
+                        int ti = (timingsIndex + i) % (_samples * 2);
+                        copy[i] = _timings[ti];
+                    }
+                }
+                //calculate percentiles
+                PercentileStats.Enqueue(_statsCalculator(copy));
+            }
+        }
+
+        public static TimingPercentileStats CalculateSomePercentiles(long[] timingArr)
+        {
+            var stats = new TimingPercentileStats()
+            {
+                Time = DateTime.Now,
+                Centile = new int[] { -50, 1 },
```

Comment:
No change needed for this PR, but consider comments explaining these magic numbers (-50 and 1), and why AvgTiming has 2 elements.